### PR TITLE
bpo-46076: Improve documentation for per-attribute docstrings with `__slots__`

### PR DESCRIPTION
**Pull request by AlexWaygood**
https://bugs.python.org/issue46076

### CODE REVIEW

**Comment by TeamSpen210** [comment]
The [pydoc](https://github.com/python/cpython/blob/9f8f45144b6f0ad481e80570538cce89b414f7f9/Lib/pydoc.py#L141) code currently checks specifically for dicts, so perhaps the docs should specify only those are expected (or that code changed)?

**Comment by AlexWaygood** [comment]
> The [pydoc](https://github.com/python/cpython/blob/9f8f45144b6f0ad481e80570538cce89b414f7f9/Lib/pydoc.py#L141) code currently checks specifically for dicts

Good catch! Whether the pydoc documentation should be changed is arguable, but I think that discussion should happen in another issue — for now, I'll just update this patch to note that only `dict` is supported.

**Comment by AlexWaygood** [comment]
@stevendaprano, you've been heavily involved in the [recent python-ideas thread](https://mail.python.org/archives/list/python-ideas@python.org/thread/VV5MOSIRVSTRDTA5NFVVUXGD2JFBERRS/) on this topic — any thoughts on this PR? 🙂

**Comment by AlexWaygood** [comment]
One thing I'm unsure of: is there anywhere else that this feature should be noted? I couldn't find a "docstring tutorial" where a mention of this feature would easily fit in.

(If there are any other places where a mention would be good, I think a link to the data model would suffice — I'm not suggesting duplicating documentation of this feature in multiple places.)

**Review by sobolevn — APPROVED** [review]
Thanks! This was always a very useful little feature not so many people know about. 👍

**Comment by rhettinger** [comment]
The style of the reference manual is terse.  Please just make a compact one-line entry noting that \__slots__ can accept a dictionary and that the *inspect* module treats the values of that dictionary as docstring for the slot.

File: Misc/NEWS.d/next/Documentation/2021-12-14-22-27-49.bpo-46076.nYQ9a8.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Improve documentation for the ability to use a ``__slots__`` dictionary to
```
Comment (rhettinger) [inline Misc/NEWS.d/next/Documentation/2021-12-14-22-27-49.bpo-46076.nYQ9a8.rst]: Doc updates don't need a NEWS entry.

**Review by auvipy — COMMENTED** [review]
really great improvement!

**Comment by AlexWaygood** [comment]
> The style of the reference manual is terse. Please just make a compact one-line entry noting that __slots__ can accept a dictionary and that the _inspect_ module treats the values of that dictionary as docstrings for the slots.

@rhettinger, I absolutely take your point that this PR, as currently written, is out of keeping with the style of the rest of the data model documentation. However, I do think the example is helpful, and this feature is not currently documented anywhere else.

If I pare this PR down to a single sentence, as you suggest, would you be open to a proposal (in a separate PR) to add a mention of this feature in the [Descriptor HowTo](https://docs.python.org/3/howto/descriptor.html#member-objects-and-slots)? I can't think of anywhere else where an example of this sort would fit well; `__slots__` are only described in detail in the data model and the Descriptor HowTo.

File: Doc/reference/datamodel.rst [hunk 2]
```diff
@@ -1962,6 +1960,53 @@ Notes on using *__slots__*
...
+   |  Data descriptors defined here:
+   |
+   |  rank
+   |      A positive integer in the range 2 <= x <= 14
+   |
+   |  suit
+   |      Either "Spades", "Hearts", "Clubs" or "Diamonds"
+
+
+.. versionadded:: 3.8
```
Comment (sobolevn) [inline Doc/reference/datamodel.rst]: One more note: it was not clear to me which feature was added in 3.8 on the first sight. Firstly, I thought about `__slots__ = {}` notation. Not about `help()` support.

I even had to test this to make sure mypy works correctly for this case 😆

Comment (AlexWaygood) [reply Doc/reference/datamodel.rst]: Should be fixed now 👍

**Comment by stevendaprano** [comment]
Thanks Alex, this looks good to me. And thanks Raymond for doing the 

work in the first place all those releases back.

File: Doc/reference/datamodel.rst [hunk 3]
```diff
@@ -1962,6 +1960,55 @@ Notes on using *__slots__*
   of the iterator's values. However, the *__slots__* attribute will be an empty
   iterator.
 
+Per-attribute docstrings with *__slots__*
```
Comment (rhettinger) [inline Doc/reference/datamodel.rst]: This is too extensive for the reference manual.  I would omit this section entirely.

Instead, focus on correcting lines 1947-1949, perhaps like this, "If a mapping is used, the mapping keys will be used as the slot names, and the *inspect* module will treat the mapping values as attribute docstrings which will be shown in *help()*.

**Review by rhettinger — CHANGES_REQUESTED** [review]

**Comment by rhettinger** [comment]
Several thoughts:

* The descriptor how-to is primarily about descriptors.  This is unrelated.
* This large example doesn't belong anywhere in the docs.
* This is a really minor feature.  
* We use it only once in the standard library and it isn't visible because the attributes are private.
* IMO it doesn't even warrant a tutorial section (we only give a couple lines to enumerate() or zip().
* The feature isn't universal. For example there isn't a way to use it with dataclasses.
* The docstring feature for named tuples only warranted a couple of lines.
* This really isn't even a feature of '\__slots__`.  Rather it is a capability that was added to *inspect*.
* The docs not there for "promoting" features.  A blog post is better medium for that.
* Please make the requested changes and make a small note in the reference manual.

**Comment by AlexWaygood** [comment]
Thanks for the feedback, Raymond -- I appreciate you taking the time! I have made the requested changes; please review again.

**Review by rhettinger — APPROVED** [review]

**Comment by rhettinger** [comment]
Thanks for the PR.

**Pull request merged**